### DISCUSSION TITLE
GitHub Actions: automatically push NuGet package

### DISCUSSION THREAD

**Discussion by PieterjanDeClippel** [community/community#26906] [Actions]
<aside class="onebox stackexchange">
  <header class="source">
      <a href="https://stackoverflow.com/questions/63489418/github-actions-automatically-push-nuget-package" target="_blank" rel="nofollow noopener">stackoverflow.com</a>
  </header>
  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/8941307/pieterjan" target="_blank" rel="nofollow noopener">
    <img alt="Pieterjan" src="https://user-images.githubusercontent.com/9629574/181100543-2ef0f0ff-877e-4af2-912f-79fca5df2ffe.png" class="thumbnail onebox-avatar" width="128" height="128">
  </a>
<h4>
  <a href="https://stackoverflow.com/questions/63489418/github-actions-automatically-push-nuget-package" target="_blank" rel="nofollow noopener">GitHub Actions: automatically push NuGet package</a>
</h4>

<div class="tags">
  <strong>nuget-package, github-actions, github-package-registry</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/8941307/pieterjan" target="_blank" rel="nofollow noopener">
    Pieterjan
  </a>
  on <a href="https://stackoverflow.com/questions/63489418/github-actions-automatically-push-nuget-package" target="_blank" rel="nofollow noopener">02:39PM - 19 Aug 20 UTC</a>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I’m trying to configure my Github repository in order to automatically have a NuGet package built and pushed to both <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> and <a href="http://github.com" rel="nofollow noopener">github.com</a>. So what I want is that each time a commit is made on the master branch, or another branch is merged into the master, github publishes a new Nuget package of the head of the master to both Nuget and Github.</p>
<h2>NuGet</h2>
<ol>
<li>On my nuget organization account, I generated an access token (username - API keys - Create)</li>
<li>On Github (select your organization - View organization - Settings tab - Secrets) I added a secret with the name <strong>PUBLISH_TO_NUGET_ORG</strong> and my nuget access token</li>
</ol>
<h2>Github</h2>
<ol>
<li>On my personal account, I generated an access token (Account - Settings - Developer settings - Personal access tokens - generate)</li>
<li>On Github I added a secret with the name <strong>PUBLISH_TO_GITHUB_COM</strong> and my github access token</li>
</ol>
<p>These are the scopes for my Github access token:</p>
<p><a href="https://i.stack.imgur.com/aR7wU.png" rel="nofollow noopener"><img src="https://i.stack.imgur.com/aR7wU.png" alt="Scopes for my Github access token" width="540" height="321"></a></p>
<h2>Setup</h2>
<p>In my github repository I’ve setup an action to restore, build, test, pack and publish:</p>
<pre><code>name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1
      with:
        dotnet-version: 3.1.301
    - name: Install dependencies
      run: dotnet restore
    - name: Build
      run: dotnet build --configuration Release --no-restore
    - name: Test
      run: dotnet test --no-restore --verbosity normal
#    - name: Publish
#      uses: brandedoutcast/publish-nuget@v2.5.2
#      with:
#          PROJECT_FILE_PATH: MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.csproj
#          NUGET_KEY: ${{secrets.PUBLISH_TO_NUGET_ORG}}
#          INCLUDE_SYMBOLS: true
    - name: Pack
      run: dotnet pack --no-build --configuration Release MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.csproj --output .
    - name: PushNuget
      run: dotnet nuget push *.nupkg --source https://api.nuget.org/v3/index.json --api-key ${{secrets.PUBLISH_TO_NUGET_ORG}} --skip-duplicate
    - name: AddGithubSource
      run: dotnet nuget add source --username PieterjanDeClippel --password ${{secrets.PUBLISH_TO_GITHUB_COM}} --name github https://nuget.pkg.github.com/MintPlayer/index.json
    - name: PushGithub
      run: dotnet nuget push *.nupkg --source github --skip-duplicate
</code></pre>
<p>The push to <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> works fine, but the push to my GitHub feed fails with an Unauthorized error.</p>
<p>I’ve taken a look at some plugins <a href="https://github.com/brandedoutcast/publish-nuget/blob/c12b8546b67672ee38ac87bea491ac94a587f7cc/index.js#L67" rel="nofollow noopener">like this one</a>, and I want to embed this into my action in order not to build my project multiple times.</p>
<p>First take:</p>
<pre><code>dotnet nuget push *.nupkg --source https://nuget.pkg.github.com/MintPlayer/index.json --api-key ${{secrets.PUBLISH_TO_GITHUB_COM}} --skip-duplicate
</code></pre>
<p>Result:<br>
<a href="https://i.stack.imgur.com/fdl0W.png" rel="nofollow noopener"><img src="https://i.stack.imgur.com/fdl0W.png" alt="warn : Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured." width="690" height="99"></a></p>
<pre><code>warn : Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured.
</code></pre>
<p>Second take with multiple commands:</p>
<pre><code>dotnet nuget add source --username PieterjanDeClippel --password ${{secrets.PUBLISH_TO_GITHUB_COM}} --name github https://nuget.pkg.github.com/MintPlayer/index.json
dotnet nuget push *.nupkg --source github --skip-duplicate
</code></pre>
<p>This one fails with the following (obvious) message:</p>
<pre><code>error: Password encryption is not supported on .NET Core for this platform. The following feed try to use an encrypted password: 'github'. You can use a clear text password as a workaround.
error:   Encryption is not supported on non-Windows platforms.
</code></pre>
<p>Does anyone have any experience with automated publishing of Nuget packages to Github?</p>
<p><a href="https://github.com/MintPlayer/MintPlayer.SeasonChecker/blob/694924bf1cd60f68a3214734b734ae1f7cb5282a/.github/workflows/dotnet-core.yml" rel="nofollow noopener">Link to action configuration file</a></p>
<h2>Edit</h2>
<p>I tried sending a POST request:</p>
<ul>
<li>Url: <a href="https://api.github.com/user" rel="nofollow noopener">https://api.github.com/user</a>
</li>
<li>Authorization: Basic Auth</li>
<li>Username:</li>
<li>Password: <code>&lt;my-api-key&gt;</code>
</li>
</ul>
<p>And I’m getting my user information, so my access token definitely works.</p>

**Comment by PieterjanDeClippel** [comment]
<p>Also when I visit <a href="https://nuget.pkg.github.com/MintPlayer/MintPlayer.SeasonChecker/index.json" rel="nofollow noopener">https://nuget.pkg.github.com/MintPlayer/MintPlayer.SeasonChecker/index.json</a> in my webbrowser, I get a login prompt and after signing in with a PAT (and not my password) I get the information for my NuGet package. So my PAT definitely works.</p>
<p>Does anybody have an idea how to solve this?</p>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/pieterjandeclippel">@PieterjanDeClippel</a>,</p>
<p>Instead of using, <code>dotnet nuget add source</code>, you can use the <code>setup-dotnet</code> to create the appropriate <code>nuget.config</code> file. For example"</p>
<pre><code class="lang-auto">    - uses: actions/setup-dotnet@v1.5.0
      with:
        dotnet-version: '3.1.x' # SDK Version to use.
        source-url: https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json
      env:
        NUGET_AUTH_TOKEN: '%NUGET_AUTH_TOKEN%'

    - name: Publish the package to GPR
      run: dotnet nuget push path/to/*.nupkg
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>This should work to associate your package with the same repository as the workflow. You only need to create a PAT if you’re associating it with a different repository.</p>
<p>I hope that helps!</p>

**Comment by PieterjanDeClippel** [comment]
<p>Hi,</p>
<p>I’ll try that. So if I understand correctly, the <code>github.token</code> variable is pulled from the action, and I don’t have to define it myself.</p>

**Comment by PieterjanDeClippel** [comment]
<p>Sweet, It’s working now.</p>
<pre><code class="lang-auto">name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1.5.0
      with:
        dotnet-version: 3.1.301
        source-url: https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json
      env:
        NUGET_AUTH_TOKEN: '%NUGET_AUTH_TOKEN%'
    - name: Install dependencies
      run: dotnet restore
    - name: Build
      run: dotnet build --configuration Release --no-restore
    - name: Test
      run: dotnet test --no-restore --verbosity normal
    - name: Pack
      run: dotnet pack --no-build --configuration Release MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.csproj --output .
    - name: PushNuget
      run: dotnet nuget push *.nupkg --source https://api.nuget.org/v3/index.json --api-key ${{secrets.PUBLISH_TO_NUGET_ORG}} --skip-duplicate
    - name: PushGithub
      run: dotnet nuget push *.nupkg --no-symbols --skip-duplicate
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>Very nice implementation.</p>
<p>Thanks for the help.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="PieterjanDeClippel" data-post="4" data-topic="128242">
<div class="title">
<div class="quote-controls"></div>
 PieterjanDeClippel:</div>
<blockquote>
<p>I’ll try that. So if I understand correctly, the <code>github.token</code> variable is pulled from the action, and I don’t have to define it myself.</p>
</blockquote>
</aside>
<p>The <code>${{ github.token }}</code> is automatically created by the workflow and has write access to the home repository.</p>
<p>I’m glad it’s now working! Let me know if you notice any other issues.</p>

**Comment by PieterjanDeClippel** [comment]
<p>It still doesn’t seem to be working correctly.</p>
<p>GitHub succeeded in pushing the package the first attempt. But when I made another change + version increment, the Action fails again, this time at the <code>Install dependencies</code> stage.</p>
<p>It appears that dotnet is trying to restore <strong>ALL</strong> packages from my own NuGet source now, instead of the source where they are hosted.</p>
<pre><code class="lang-auto">Determining projects to restore...
/opt/hostedtoolcache/dncs/3.1.301/x64/sdk/3.1.301/NuGet.targets(128,5): warning : Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured. [/home/runner/work/MintPlayer.SeasonChecker/MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.sln]
Retrying 'FindPackagesByIdAsync' for source 'https://nuget.pkg.github.com/MintPlayer/download/netstandard.library/index.json'.
Response status code does not indicate success: 401 (Unauthorized).
/opt/hostedtoolcache/dncs/3.1.301/x64/sdk/3.1.301/NuGet.targets(128,5): warning : Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured. [/home/runner/work/MintPlayer.SeasonChecker/MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.sln]
</code></pre>
<p>So dotnet looks for <code>NetStandard.Library</code> in my own NuGet feed.</p>
<p>How can I fix this?</p>

**Comment by jcansdale** [comment]
<p>What does your <code>nuget.config</code> file look like?</p>

**Comment by PieterjanDeClippel** [comment]
<p>Right now there’s no <code>nuget.config</code> in my solution/project.</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/MintPlayer/MintPlayer.SeasonChecker" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/9629574/181100554-aa6012ca-c454-4c16-9b92-d4b39ff611e4.png" class="thumbnail onebox-avatar" width="192" height="192">

<h3><a href="https://github.com/MintPlayer/MintPlayer.SeasonChecker" target="_blank" rel="nofollow noopener">MintPlayer/MintPlayer.SeasonChecker</a></h3>

<p>Find out the season for a date. Contribute to MintPlayer/MintPlayer.SeasonChecker development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Are you aiming at defining my package sources for the action? Right now on my computer I’m using the system-wide package sources. So when I run</p>
<pre><code>dotnet nuget list source
</code></pre>
<p>command from anywhere the list of sources is shown (including my own nuget feed).</p>
<p>Should I create <a href="https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#packagesources" rel="nofollow noopener">a nuget.config like this</a> containing my own nuget feed, and remove the <code>source-url</code> from the action?</p>

**Comment by PieterjanDeClippel** [comment]
<p>Okay that was it.</p>
<p>I added a nuget.config to my solution folder</p>
<pre><code class="lang-auto">&lt;?xml version="1.0" encoding="utf-8"?&gt;
&lt;configuration&gt;
  &lt;packageSources&gt;
    &lt;add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" /&gt;
  &lt;/packageSources&gt;
&lt;/configuration&gt;
</code></pre>
<p>And now my deployment is working fine.</p>
<p>Thanks a lot for the help</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/pieterjandeclippel">@PieterjanDeClippel</a>,</p>
<aside class="quote no-group" data-username="PieterjanDeClippel" data-post="7" data-topic="128242">
<div class="title">
<div class="quote-controls"></div>
 PieterjanDeClippel:</div>
<blockquote>
<p>It appears that dotnet is trying to restore <strong>ALL</strong> packages from my own NuGet source now, instead of the source where they are hosted.</p>
</blockquote>
</aside>
<p>I think it was trying to restore from both, but GitHub Packages was failing because <code>NUGET_AUTH_TOKEN</code> wasn’t being passed.</p>
<p>Could you try changing the <code>dotnet restore</code> step to this and remove <code>nuget.config</code>?</p>
<pre><code class="lang-auto">    - name: Install dependencies
      run: dotnet restore
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>BTW, it is possible to pass the token directly to <code>actions/setup-dotnet</code>, but I prefer to only pass the token to steps that need it. That way you’re only trusting <code>dotnet restore</code> and <code>dotnet nuget</code> with your token.</p>

**Comment by joelrevans** [comment]
<p>Piggybacking off of this. I have implemented a near identical solution that pushes multiple packages, but the --skip-duplicate flag is ignored while pushing and causes a failure. Is there a reason why that flag would be ignored?</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/ecc8ee603340dc07b09e212c96363e123a686672.png" data-download-href="/uploads/short-url/xMHaUHv35pwDPKdok5F2jawQO7o.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/10842234/181100555-f4b156ce-f88b-4e2e-99c1-7b24ddbe28ec.png" alt="image" data-base62-sha1="xMHaUHv35pwDPKdok5F2jawQO7o" width="577" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/ecc8ee603340dc07b09e212c96363e123a686672_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">830×719 45.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Reply by PieterjanDeClippel** [reply]
Hi, it seems this was fixed since .NET 7. For pushing to GPR, you no longer need the MSBuild SDK and `nuget.exe push **/*.nupkg`, so you can remove it and replace the step command with `dotnet nuget push **/*.nupkg --skip-duplicate` which will effectively work as expected

**Comment by PieterjanDeClippel** [comment]
<p><a class="mention" href="/joelrevans">@joelrevans</a> I switched from</p>
<pre><code>dotnet nuget push *.nupkg --no-symbols --skip-duplicate
</code></pre>
<p>to</p>
<pre><code>nuget.exe push "**/*.nupkg" -NoSymbols -SkipDuplicate
</code></pre>
<p>And now the step does not error when attemting to push a duplicate package.</p>
<p><img src="https://user-images.githubusercontent.com/9629574/181100558-4344640b-57c1-4581-b509-18c0d994ff6c.png" alt="image" data-base62-sha1="wYauYKSRnznmC7QLmHAUkPpNaSi" width="225" height="370"></p>
<p>I needed to switch to nuget.exe <a href="https://github.community/t/push-nuget-package-to-github-com-the-response-ended-prematurely-an-error-occurred-while-sending-the-request/131032">because of this</a></p>
